### PR TITLE
[6/n][FSDP] Update _sharded_pre_load_state_dict_hook to use DTensor when use_dtensor=True in ShardedStateDictConfig 

### PR DESCRIPTION
**Pull request by wz337**
This allows us use use_dtensor=True for ShardedStateDictConfig() before calling model.load_state_dict(). It only works for offload_to_cpu=False now. 

Next PR will make use_dtensor=True work with offload_to_cpu=True for load_state_dict().

 

cc @zhaojuanmao @mrshenli @rohan-varma @awgu

### CODE REVIEW

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 1]
```diff
@@ -599,55 +590,63 @@ def _sharded_pre_load_state_dict_hook(
...
+                dist.all_gather_into_tensor(
+                    tensor, local_tensor, group=fsdp_state.process_group
+                )
+            tensor = tensor.narrow(0, 0, param_numel).reshape(param.size())
+            state_dict[fqn_from_global_root] = tensor
         else:
-            dist.all_gather_into_tensor(
-                tensor, local_tensor, group=fsdp_state.process_group
+            # TODO: make use_dtensor=True work with offload_to_cpu=True.
+            param = param.redistribute(
```
Comment (fegin) [inline torch/distributed/fsdp/_state_dict_utils.py:646]: If param is in CPU, will this function still work? We need to consider parameters in `state_dict` are loaded to CPU first before calling `load_state_dict()`.

Comment (wz337) [reply torch/distributed/fsdp/_state_dict_utils.py:654]: > If param is in CPU, will this function still work? We need to consider parameters in `state_dict` are loaded to CPU first before calling `load_state_dict()`.

Good catch. Thanks for raising this! Updated to handle this case and included test cases to cover this.

File: test/distributed/fsdp/test_fsdp_dtensor_state_dict.py [hunk 2]
```diff
@@ -129,6 +132,50 @@ def test_dtensor_sharded_model_state_dict(self, offload_to_cpu):
...
+                use_dtensor=True,
+                offload_to_cpu=False,
+            ),
+        )
+
+        checkpoint = io.BytesIO()
+        torch.save(model.state_dict(), checkpoint)
+        # Deepcopy to save current state_dict to compare with the loaded state dict below.
+        ref_state_dict = deepcopy(model.state_dict())
+        # Hmmm...Why does ```torch.save(ref_state_dict, checkpoint)``` not work while torch.save(model.state_dict()) work?
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:155]: This is pretty strange.

File: test/distributed/fsdp/test_fsdp_dtensor_state_dict.py [hunk 3]
```diff
@@ -129,6 +132,50 @@ def test_dtensor_sharded_model_state_dict(self, offload_to_cpu):
...
+        checkpoint.seek(0)
+        load_ref_state_dict = torch.load(checkpoint, map_location=map_location)
+        model.load_state_dict(load_ref_state_dict)
+        new_state_dict = model.state_dict()
+
+        # Check whether new_state_dict is the same as ref_state_dict.
+        for ref_state_dict, new_state_dict in zip(
+            ref_state_dict.items(), new_state_dict.items()
+        ):
+            k1, v1 = ref_state_dict
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:171]: this is a little confusing, we're calling `ref_state_dict` the (key, value) tuple, but `ref_state_dict` is also the name of the dict. Can we just do:

`for (k1, v1), (k2, v2) in zip(ref_state_dict.items, ...)`

File: test/distributed/fsdp/test_fsdp_dtensor_state_dict.py [hunk 4]
```diff
@@ -129,6 +132,50 @@ def test_dtensor_sharded_model_state_dict(self, offload_to_cpu):
...
+            k2, v2 = new_state_dict
+            self.assertEqual(k1, k2)
+            # check whether local_tensor are the same
+            self.assertEqual(v1.to_local(), v2.to_local())
+            # check whether device are the same
+            self.assertEqual(v1.to_local().device, v2.to_local().device)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:177]: Does DTensor support a __equals__ that we can use here? Should we validate things like placements, sharding spec, etc?

File: test/distributed/fsdp/test_fsdp_dtensor_state_dict.py [hunk 5]
```diff
@@ -129,6 +132,50 @@ def test_dtensor_sharded_model_state_dict(self, offload_to_cpu):
...
+        # Hmmm...Why does ```torch.save(ref_state_dict, checkpoint)``` not work while torch.save(model.state_dict()) work?
+
+        # Update the parameters so model.state_dict() will be different from ref_dtensor_sd.
+        model(model.get_input()).sum().backward()
+        optim.step()
+
+        # # Load ref_state_dict back.
+        checkpoint.seek(0)
+        load_ref_state_dict = torch.load(checkpoint, map_location=map_location)
+        model.load_state_dict(load_ref_state_dict)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_dtensor_state_dict.py:164]: do we need to test model is on CPU?

**Review by rohan-varma — COMMENTED** [review]
LGTM, some comments on testing. will let @fegin to stamp

**Review by fegin — APPROVED** [review]

**Pull request closed without merging**